### PR TITLE
Add `NON_GCP_PRIVATE_IP_PORT` as networkEndpointType for NetworkEndpointGroup

### PR DESCRIPTION
**Pull request by AlexanderEllis**
This change adds `NON_GCP_PRIVATE_IP_PORT` as a `network_endpoint_type` option for the `google_compute_network_endpoint_group` resource.  This is in support for Hybrid Load Balancing.

I added two examples: one with a basic `google_compute_network_endpoint_group` with the new type, and one with a global external managed load balancer with hybrid connectivity (roughly following the configuration at https://cloud.google.com/load-balancing/docs/https/setting-up-ext-https-hybrid). I also added documentation about the valid combination of this new type and backend service load balancing scheme/balancing modes.

Part of https://github.com/hashicorp/terraform-provider-google/issues/10040

For posterity, this change also sets the `instance` attribute on the `google_compute_network_endpoint` resource to be optional, as Hybrid connectivity NEGs use endpoints with just IP and PORT (without an instance). See https://cloud.google.com/load-balancing/docs/negs#hybrid-neg and https://cloud.google.com/load-balancing/docs/https/setting-up-ext-global-https-hybrid#set_up_the_hybrid_connectivity_neg for more details.

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
compute: add `NON_GCP_PRIVATE_IP_PORT` value for `network_endpoint_type` in the `google_compute_network_endpoint_group` resource
```

```release-note:enhancement
compute: Update `instance` attribute for `google_compute_network_endpoint` to be optional, as Hybrid connectivity NEGs use network endpoints with just IP and Port.
```

### CODE REVIEW

**Comment by melinath** [comment]
b/202539913, and more context: b/202525233, b/181008964

File: mmv1/products/compute/terraform.yaml [hunk 1]
```diff
@@ -1498,6 +1513,12 @@ overrides: !ruby/object:Overrides::ResourceOverrides
           neg_name: "my-lb-neg"
           network_name: "neg-network"
           subnetwork_name: "neg-subnetwork"
+      - !ruby/object:Provider::Terraform::Examples
+        name: "network_endpoint_group_non_gcp"
+        primary_resource_id: "neg"
+        vars:
+          neg_name: "my-lb-neg"
+          network_name: "neg-network"
```
Comment (melinath) [inline mmv1/products/compute/terraform.yaml:1521]: It looks like `TestAccComputeNetworkEndpointGroup_networkEndpointGroupNonGcpExample` is failing with the following error:

```
Error: "name" ("") doesn't match regexp "^(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?)$"
on terraform_plugin_test.tf line 17, in resource "google_compute_subnetwork" "default":
17:   name          = ""
```

I think you may need to add a `subnetwork_name` variable value here.

File: mmv1/products/compute/api.yaml [hunk 2]
```diff
@@ -8155,9 +8155,16 @@ objects:
       - !ruby/object:Api::Type::Enum
         name: 'networkEndpointType'
         description: |
-          Type of network endpoints in this network endpoint group.
+          Type of network endpoints in this network endpoint group. 
+          NON_GCP_PRIVATE_IP_PORT is used for hybrid connectivity network 
+          endpoint groups (see https://cloud.google.com/load-balancing/docs/hybrid).
+          Note that NON_GCP_PRIVATE_IP_PORT can only be used with Backend Services
+          that 1) have the following load balancing schemes: EXTERNAL, EXTERNAL_MANAGED,
+          INTERNAL_MANAGED, and INTERNAL_SELF_MANAGED and 2) to support the RATE or
```
Comment (melinath) [inline mmv1/products/compute/api.yaml:8163]:
```suggestion
          INTERNAL_MANAGED, and INTERNAL_SELF_MANAGED and 2) support the RATE or 
```

**Review by melinath — CHANGES_REQUESTED** [review]
This looks good overall - it looks like one of the new tests is failing. I've noted a potential fix below. If you can test it locally prior to pushing, that would be great!

File: mmv1/products/compute/api.yaml (hunk 2 above)
Comment (AlexanderEllis) [reply mmv1/products/compute/api.yaml]: Good catch, thanks!

File: mmv1/products/compute/terraform.yaml (hunk 1 above)
Comment (AlexanderEllis) [reply mmv1/products/compute/terraform.yaml:1521]: Whoops great catch! In all of my forwarding rule excitement, I forgot to go back and fix this test. It actually was a little out of date - subnetworks aren't allowed with non-GCP, so these examples shouldn't have a subnetwork (I updated that for the forwarding rule, but not for this one).

Updated the `network_endpoint_group_non_gcp.tf.erb` file to remove the subnetwork, generated the provider, and ran it:

image

Thanks again!

**Comment by melinath** [comment]
The CI/CD tests failed but it looks like for unrelated reasons. I've started a run of just the new tests here: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/256720

**Comment by melinath** [comment]
also gonna do another /gcbrun for good measure

**Review by melinath — APPROVED** [review]
LGTM - I'll merge in the morning assuming the gcbrun looks good.

**Comment by ct-dh** [comment]
I was working on my own version of this but hadn't got round to tidying it up to raise a PR. Does this PR support importing the hybrid NEGs as is? I found I had to make more changes than in this PR to get imports to work correctly (this is the diff I had, but as I say I haven't had a chance to tidy it up, please feel free to grab anything if needed): https://github.com/GoogleCloudPlatform/magic-modules/compare/master...ct-dh:feature/hybrid-neg-support?expand=1

**Review by melinath — CHANGES_REQUESTED** [review]
Per discussion above, sounds like there are some changes to be made.

**Comment by AlexanderEllis** [comment]
Updated NetworkEndpoint instance param to be optional and added an example/tests exercising it. Followed the patterns in `resource_compute_global_network_endpoint` examples for the optional params (custom import and handling). Hat tip to @ct-dh again for the suggestion!

image

**Comment by melinath** [comment]
New test run for just the new tests: https://ci-oss.hashicorp.engineering/buildConfiguration/GoogleCloudBeta_ProviderGoogleCloudBetaMmUpstream/257226

**Comment by melinath** [comment]
Trying /gcbrun again

**Comment by melinath** [comment]
@AlexanderEllis quick question - could you add a note explaining why the `instance` field is being changed to optional?

**Comment by AlexanderEllis** [comment]
> @AlexanderEllis quick question - could you add a note explaining why the `instance` field is being changed to optional?

Done: added separate release note and additional links in description for posterity. Thanks!

**Review by melinath — APPROVED** [review]
LGTM - remaining test failures are unrelated.

**Pull request merged**